### PR TITLE
Add VIP Search widget

### PR DESCRIPTION
**Pull request by pauarge**
## Description

This PR introduces a WordPress widget that can be used as a drop-in UI element for VIP Enterprise Search. It is based in Jetpack's Search Widget, but this first iteration doesn't include ES facets (tags).

### Add VIP Search Widget

Adding a first iteration of a Search Widget for VIP Search.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples.

## Steps to Test

1. Check out PR.
2. Make sure VIP Search is enabled.
3. Set up a Search Widget in wp-admin.
4. Perform a search in the FE.

### CODE REVIEW

File: search/ui/class-vip-search-widget.php [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+	/**
+	 * Default sort order for search results.
+	 *
+	 * @var string
+	 */
+	const DEFAULT_SORT = 'relevance_desc';
+
+	public function __construct( $name = null ) {
+		if ( empty( $name ) ) {
+			$name = 'Search (VIP Enterprise)';
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:22]:
```suggestion
			$name = 'Enterprise Search';
```

File: search/ui/class-vip-search-widget.php [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+		}
+
+		parent::__construct(
+			'vip-search-widget',
+			$name,
+			array( 'description' => 'UI for VIP Enterprise Search' )
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:28]:
```suggestion
			array( 'description' => 'Enterprise Search' )
```

File: search/ui/class-vip-search-widget.php [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+
+		wp_enqueue_script(
+			'vip-search-widget',
+			plugins_url( 'js/search-widget.js', __FILE__ ),
+			array(),
+			1.0,
+			true
+		);
+
+		wp_enqueue_style( 'vip-search-widget', plugins_url( 'css/search-widget-frontend.css', __FILE__ ), array(), 1.0 );
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:53]: Is there any particular reason the JS file is named as `search-widget.js` but the CSS one is `search-widget-frontend.css`?

File: search/ui/class-vip-search-widget.php [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+		$instance = $this->vip_search_populate_defaults( $instance );
+
+		if ( empty( $instance['search_box_enabled'] ) && empty( $instance['user_sort_enabled'] ) ) {
+			return;
+		}
+
+		$title = ! empty( $instance['title'] ) ? $instance['title'] : '';
+
+		/** This filter is documented in core/src/wp-includes/default-widgets.php */
+		$title = apply_filters( 'widget_title', $title, $instance, $this->id_base );
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:101]: What would be the use-case of utilizing this core filter?

File: search/ui/class-vip-search-widget.php [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+		// it separately when it's not.
+		if ( ! empty( $instance['search_box_enabled'] ) ) {
+			$this->render_widget_search_form( $instance['post_types'], $orderby, $order );
+		}
+
+		if ( ! empty( $instance['search_box_enabled'] ) && ! empty( $instance['user_sort_enabled'] ) ) :
+			?>
+			<div class="vip-search-sort-wrapper">
+				<label>
+					<?php esc_html_e( 'Sort by', 'jetpack' ); ?>
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:126]: I know we technically don't have a separate text domain, but seems we should fix this copy-pasta
```suggestion
					<?php esc_html_e( 'Sort by', 'vip-search' ); ?>
```

File: search/ui/class-vip-search-widget.php [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+	 * @param array  $instance The current widget instance.
+	 * @param string $order    The order to initialize the select with.
+	 * @param string $orderby  The orderby to initialize the select with.
+	 *
+	 */
+	private function maybe_render_sort_javascript( array $instance, string $order, string $orderby ) {
+		if ( ! empty( $instance['user_sort_enabled'] ) ) :
+			?>
+			<script type="text/javascript">
+				var vipSearchModuleSorting = function() {
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:163]: What do you think about moving this to a JS file and only printing the required "dynamic" portion of the data and pass that as function arguments?

File: search/ui/class-vip-search-widget.php [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+		$classes = sprintf(
+			'vip-search-filters-widget %s %s',
+			$instance['search_box_enabled'] ? '' : 'hide-post-types',
+			$this->id
+		);
+		?>
+		<div class="<?php echo esc_attr( $classes ); ?>">
+			<p>
+				<label for="<?php echo esc_attr( $this->get_field_id( 'title' ) ); ?>">
+					<?php esc_html_e( 'Title (optional):', 'jetpack' ); ?>
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:270]:
```suggestion
					<?php esc_html_e( 'Title (optional):', 'vip-search' ); ?>
```

File: search/ui/class-vip-search-widget.php [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+
+			<p>
+				<label>
+					<input
+						type="checkbox"
+						class="vip-search-filters-widget__search-box-enabled"
+						name="<?php echo esc_attr( $this->get_field_name( 'search_box_enabled' ) ); ?>"
+						<?php checked( $instance['search_box_enabled'] ); ?>
+					/>
+					<?php esc_html_e( 'Show search box', 'jetpack' ); ?>
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:289]:
```suggestion
					<?php esc_html_e( 'Show search box', 'vip-search' ); ?>
```

File: search/ui/class-vip-search-widget.php [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+			<p>
+				<label>
+					<input
+						type="checkbox"
+						class="vip-search-filters-widget__sort-controls-enabled"
+						name="<?php echo esc_attr( $this->get_field_name( 'user_sort_enabled' ) ); ?>"
+						<?php checked( $instance['user_sort_enabled'] ); ?>
+						<?php disabled( ! $instance['search_box_enabled'] ); ?>
+					/>
+					<?php esc_html_e( 'Show sort selection dropdown', 'jetpack' ); ?>
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:302]:
```suggestion
					<?php esc_html_e( 'Show sort selection dropdown', 'vip-search' ); ?>
```

File: search/ui/class-vip-search-widget.php [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+				</label>
+			</p>
+
+			<p class="vip-search-filters-widget__post-types-select">
+				<label><?php esc_html_e( 'Post types to search (minimum of 1):', 'jetpack' ); ?></label>
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:307]:
```suggestion
				<label><?php esc_html_e( 'Post types to search (minimum of 1):', 'vip-search' ); ?></label>
```

File: search/ui/class-vip-search-widget.php [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+							<?php checked( empty( $instance['post_types'] ) || in_array( $post_type->name, $instance['post_types'] ) ); ?>
+						/>&nbsp;
+						<?php echo esc_html( $post_type->label ); ?>
+					</label>
+				<?php endforeach; ?>
+			</p>
+
+			<p>
+				<label>
+					<?php esc_html_e( 'Default sort order:', 'jetpack' ); ?>
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:323]:
```suggestion
					<?php esc_html_e( 'Default sort order:', 'vip-search' ); ?>
```

File: search/ui/class-vip-search-widget.php [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+		// we need to swap the array order though.
+		if ( count( $array_1 ) !== count( $array_2 ) && count( $array_2 ) > count( $array_1 ) ) {
+			$temp    = $array_1;
+			$array_1 = $array_2;
+			$array_2 = $temp;
+		}
+
+		// Disregard keys
+		return array_values( array_diff( $array_1, $array_2 ) );
+	}
```
Comment (rinatkhaziev) [inline search/ui/class-vip-search-widget.php:440]: Are we planning to re-use this method ever?

File: search/ui/js/search-widget.js [hunk 13]
```diff
@@ -0,0 +1,47 @@
+var vipSearchModule = function () {
```
Comment (rinatkhaziev) [inline search/ui/js/search-widget.js:1]: We should use the modern JS syntax if possible. Not a big deal though.

**Review by rinatkhaziev — CHANGES_REQUESTED** [review]
Thank you! I've left some feedback, but it seems that most things are due to copy-pasting from the Jetpack Search widget code.

The biggest issue IMO is the inlined JS, I feel like it's better to put it into a separate file which will help with readability, testability, and future maintenance.

File: search/ui/class-vip-search-widget.php (hunk 12 above)
Comment (pauarge) [reply search/ui/class-vip-search-widget.php:440]: This comes from the Jetpack implementation, I'm not sure we want to use it from outside this context.

File: search/ui/js/search-widget.js (hunk 13 above)
Comment (pauarge) [reply search/ui/js/search-widget.js:1]: I agree. Again, didn't want to diverge from the Jetpack implementation at the beginning, so I think it's fine to leave it as is.

File: search/ui/class-vip-search-widget.php (hunk 3 above)
Comment (pauarge) [reply search/ui/class-vip-search-widget.php:53]: I'm just trying to stick with Jetpack's names. They have the following files (some of which we don't need yet): `search-widget-admin-ui.css`, `search-widget-frontend.css`, `search-widget.js` and `search-widget-admin.js`.

File: search/ui/class-vip-search-widget.php (hunk 4 above)
Comment (pauarge) [reply search/ui/class-vip-search-widget.php:101]: This comes from Jetpack, I guess it's to provide additional flexibility. I'm not sure how I'd use it, so I'm happy to remove it, however, I don't think it hurts.

File: search/ui/class-vip-search-widget.php (hunk 6 above)
Comment (pauarge) [reply search/ui/class-vip-search-widget.php:163]: That would be a nice improvement to have, but for now I'd rather stick to Jetpack's implementation.

**Comment by pauarge** [comment]
@rinatkhaziev Got back to you in your comments. Most of them boil down to the fact that we're taking Jetpack's implementation. Since we want to add more things that are Jetpack-based in a subsequent PR (namely filters), I'd want to stick to their code. Once we have finished this first MVP, I agree with most of the things that you say and we should definitely improve the code.

**Review by rinatkhaziev — APPROVED** [review]
Thanks for the context!

**Pull request merged**